hudi-utilities/src/main/java/org/apache/hudi/utilities/sources/helpers/IncrSourceHelper.java

```diff
@@ -78,8 +79,20 @@ public static Pair<String, Pair<String, String>> calculateBeginAndEndInstants(Ja
         "Make sure the config hoodie.deltastreamer.source.hoodieincr.num_instants is set to a positive value");
     HoodieTableMetaClient srcMetaClient = HoodieTableMetaClient.builder().setConf(jssc.hadoopConfiguration()).setBasePath(srcBasePath).setLoadActiveTimelineOnLoad(true).build();
 
-    final HoodieTimeline activeCommitTimeline =
+    // Find the earliest incomplete commit, deltacommit, or non-clustering replacecommit,
+    // so that the incremental pulls should be strictly before this instant.
+    // This is to guard around multi-writer scenarios where a commit starting later than
+    // another commit from a concurrent writer can finish earlier, leaving an inflight commit
+    // before a completed commit.
+    final Option<HoodieInstant> firstIncompleteCommit = srcMetaClient.getCommitsTimeline()
```

Comment:
for falling back to snapshot query, should we instead of "time travel query" ? L 105. why do entire snapshot and then filter out unwanted commits in the end?